File: LS_Project/static-content/sparouter/services/location-service.js
import { request, buildQuery } from './api.js';

function normalizeLocations(data) {
    if (!data) {
        return [];
    }

    if (data.locations) {
        return data.locations;
    }

    return data;
}

export const locationService = {
    getLocations: (skip = 0, limit = 10) =>
        request(`/locations?${buildQuery({ skip, limit })}`, 'GET'),

    getAllLocations: async () =>
        normalizeLocations(await locationService.getLocations(0, 100)),

    getLocationDetails: (id) =>
        request(`/locations/${id}`, 'GET'),

    getChildrenLocations: (id) =>
        request(`/locations/${id}/children`, 'GET'),

    getFullPath: (id) =>
        request(`/locations/${id}/fullpath`, 'GET'),

    createLocation: (locationData) =>
        request('/locations', 'POST', locationData)
};

File: LS_Project/src/main/kotlin/pt/isel/ls/services/LocationService.kt
package pt.isel.ls.services

import pt.isel.ls.data.StorageDataManager
import pt.isel.ls.domain.Location
import pt.isel.ls.domain.LocationPath
import pt.isel.ls.domain.dto.GetChildLocationsOutput
import pt.isel.ls.exceptions.NotFoundException
import pt.isel.ls.exceptions.UnauthorizedException
import pt.isel.ls.utils.validatePaging

class LocationService(
    dataManager: StorageDataManager,
) {
    private val locationStorage = dataManager.locationData
    private val tokenStorage = dataManager.tokenData

    fun getLocationDetails(lid: Int): Location {
        return locationStorage.getLocation(lid)
            ?: throw NotFoundException("Location with id = $lid not found")
    }

    fun getChildrenLocations(lid: Int): GetChildLocationsOutput {
        val location =
            locationStorage.getLocation(lid)
                ?: throw NotFoundException("Location with id = $lid not found")
        return GetChildLocationsOutput(locationStorage.getChildLocations(location))
    }

    fun getFullHierarchicalPath(lid: Int): LocationPath {
        locationStorage.getLocation(lid)
            ?: throw NotFoundException("Location with id = $lid not found")
        return LocationPath(locationStorage.getLocationPath(lid))
    }

    fun createNewLocation(
        location: Location,
        tokenString: String,
    ): Int {
        tokenStorage.getUserByToken(tokenString)
            ?: throw UnauthorizedException("Invalid or expired token")
        return locationStorage.createLocation(location)
    }

    fun getLocations(
        skip: Int,
        limit: Int,
    ): List<Location> {
        validatePaging(skip, limit)
        return locationStorage.getLocations(skip, limit)
    }
}
